### PR TITLE
Reordering related entities in the submission

### PR DESCRIPTION
**Pull request by LotteHofstede**
This PR adds the ability to reorder related entities during the submission.
​
It builds on #531 and it depends on https://github.com/DSpace/DSpace/pull/2610 which fixes a rest bug.

### CODE REVIEW

**Comment by benbosman** [comment]
I've updated this branch to ensure GIT will only display the commits which differ with master

**Comment by LotteHofstede** [comment]
@paulo-graca @tdonohue This PR now passes all tests and is ready to be reviewed

File: src/app/core/data/relationship.service.ts [hunk 1]
```diff
@@ -376,14 +377,39 @@ export class RelationshipService extends DataService<Relationship> {
...
+      take(1),
+    ).subscribe(() => {
+      this.removeRelationshipItemsFromCache(item1);
+      this.removeRelationshipItemsFromCache(item2);
+    });
+
+    return update$
+  }
+
+  public updatePlace(reoRel: ReorderableRelationship): Observable<RemoteData<Relationship>> {
```
Comment (tdonohue) [inline src/app/core/data/relationship.service.ts]: Please add Typedocs to this new service method.

File: src/app/shared/form/builder/ds-dynamic-form-ui/ds-dynamic-form-control-container.component.ts [hunk 2]
```diff
@@ -331,13 +343,24 @@ export class DsDynamicFormControlContainerComponent extends DynamicFormControlCo
...
   }
 
-  /**
-   * Method to remove a selected relationship from the item
-   * @param object The second item in the relationship, the submitted item being the first
-   */
-  removeSelection(object: SearchResult<Item>) {
-    this.selectableListService.deselectSingle(this.listId, object);
-    this.store.dispatch(new RemoveRelationshipAction(this.item, object.indexableObject, this.model.relationship.relationshipType))
+  moveSelection(event: CdkDragDrop<Relationship>) {
```
Comment (tdonohue) [inline src/app/shared/form/builder/ds-dynamic-form-ui/ds-dynamic-form-control-container.component.ts]: Please add TypeDocs or a basic method description

File: src/app/shared/form/builder/ds-dynamic-form-ui/existing-metadata-list-element/existing-metadata-list-element.component.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+import { ItemMetadataRepresentation } from '../../../../../core/shared/metadata-representation/item/item-metadata-representation.model';
+import { RelationshipOptions } from '../../models/relationship-options.model';
+import { RemoveRelationshipAction } from '../relation-lookup-modal/relationship.actions';
+import { SelectableListService } from '../../../../object-list/selectable-list/selectable-list.service';
+import { Store } from '@ngrx/store';
+import { AppState } from '../../../../../app.reducer';
+import { ItemSearchResult } from '../../../../object-collection/shared/item-search-result.model';
+
+// tslint:disable:max-classes-per-file
+export abstract class Reorderable {
```
Comment (tdonohue) [inline src/app/shared/form/builder/ds-dynamic-form-ui/existing-metadata-list-element/existing-metadata-list-element.component.ts]: Please add TypeDocs to this class

File: src/app/shared/form/builder/ds-dynamic-form-ui/existing-metadata-list-element/existing-metadata-list-element.component.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  constructor(public oldIndex?: number, public newIndex?: number) {
+  }
+
+  abstract getId(): string;
+
+  abstract getPlace(): number;
+}
+
+export class ReorderableRelationship extends Reorderable {
```
Comment (tdonohue) [inline src/app/shared/form/builder/ds-dynamic-form-ui/existing-metadata-list-element/existing-metadata-list-element.component.ts]: Please add TypeDocs to this class

File: src/app/shared/form/builder/ds-dynamic-form-ui/existing-metadata-list-element/existing-metadata-list-element.component.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    }
+  }
+}
+
+@Component({
+  selector: 'ds-existing-metadata-list-element',
+  templateUrl: './existing-metadata-list-element.component.html',
+  styleUrls: ['./existing-metadata-list-element.component.scss']
+})
+export class ExistingMetadataListElementComponent implements OnChanges, OnDestroy {
```
Comment (tdonohue) [inline src/app/shared/form/builder/ds-dynamic-form-ui/existing-metadata-list-element/existing-metadata-list-element.component.ts]: Please add TypeDocs to describe this Component

File: src/app/core/data/relationship.service.ts (hunk 1 above)
Comment (LotteHofstede) [reply src/app/core/data/relationship.service.ts]: Added TypeDoc

File: src/app/shared/form/builder/ds-dynamic-form-ui/existing-metadata-list-element/existing-metadata-list-element.component.ts (hunk 3 above)
Comment (LotteHofstede) [reply src/app/shared/form/builder/ds-dynamic-form-ui/existing-metadata-list-element/existing-metadata-list-element.component.ts]: Added TypeDoc

File: src/app/shared/form/builder/ds-dynamic-form-ui/existing-metadata-list-element/existing-metadata-list-element.component.ts (hunk 5 above)
Comment (LotteHofstede) [reply src/app/shared/form/builder/ds-dynamic-form-ui/existing-metadata-list-element/existing-metadata-list-element.component.ts]: Added TypeDocAdded TypeDoc

File: src/app/shared/form/builder/ds-dynamic-form-ui/existing-metadata-list-element/existing-metadata-list-element.component.ts (hunk 4 above)
Comment (LotteHofstede) [reply src/app/shared/form/builder/ds-dynamic-form-ui/existing-metadata-list-element/existing-metadata-list-element.component.ts]: Added TypeDoc

File: src/app/shared/form/builder/ds-dynamic-form-ui/ds-dynamic-form-control-container.component.ts (hunk 2 above)
Comment (LotteHofstede) [reply src/app/shared/form/builder/ds-dynamic-form-ui/ds-dynamic-form-control-container.component.ts]: Added TypeDoc

**Comment by LotteHofstede** [comment]
@tdonohue I added the requested TypeDoc, please review again when you have time

**Comment by artlowel** [comment]
Thanks for the review @paulo-graca!

>The icon for drag&drop it's very close to the remove one and it's small. I don't see it working properly if you interact with DSpace using your fingers

The way results are rendered still changes significantly between this and #541, because in that PR we also add metadata results in the mix. Currently, in that PR the drag handle is at the front and the close button is at the back (as a trash can icon). The drag icon has changed in to [this](https://fontawesome.com/icons/grip-vertical?style=solid). We'll make sure to test it with touch controls.

**Pull request merged**